Made specification of dependencies more general and removed redundant…

G-Ragghianti opened the pull request:
Three majors changes to the cmake config:

1. Generalized the specification of FFT library include and library file location.  This removes the assumption that the library is installed in the user's home directory.  Instead, the user should specify the locations using `cmake -DFIBER_FFT_INCLUDE_DIR=... -DFIBER_FFT_LIB_DIR=.. ..`.
2. Removed the `add_definitions(...)` as this seems to be unnecessary.
3. Enables the specification of the install location of FFTW and MPI using $FFTW_DIR and $MPI_DIR instead of hard-coding paths.
4. Provides the harness with the location of its own include directory since it was otherwise unable to find it unless it was located in a hard-coded directory in the $HOME.

These changes allow the code to be compiled in any location without assumptions of the location of the harness or its dependencies.  These changes are necessary for easy integration into spack.

G-Ragghianti committed:
Made specification of dependencies more general and removed redundant code

cayrols commented:
It is working on my end.

Thank you Gery for doing that.

Just as a record, when I did cmake without fftw in the PATH, the cmake did not complain.

af-ayala commented:
> Three majors changes to the cmake config:
> 
>     1. Generalized the specification of FFT library include and library file location.  This removes the assumption that the library is installed in the user's home directory.  Instead, the user should specify the locations using `cmake -DFIBER_FFT_INCLUDE_DIR=... -DFIBER_FFT_LIB_DIR=.. ..`.
> 
>     2. Removed the `add_definitions(...)` as this seems to be unnecessary.
> 
>     3. Enables the specification of the install location of FFTW and MPI using $FFTW_DIR and $MPI_DIR instead of hard-coding paths.
> 
>     4. Provides the harness with the location of its own include directory since it was otherwise unable to find it unless it was located in a hard-coded directory in the $HOME.
> 
> 
> These changes allow the code to be compiled in any location without assumptions of the location of the harness or its dependencies. These changes are necessary for easy integration into spack.

Gerald,
Thank you for your contribution, some comments:
- For the DFIBER_FFT_INCLUDE_DIR, I believe the CMake can set this by default, since the path is dependent of where Fiber is installed.
- Important: The add_definitions I put there need to stay or find a way to make it work, otherwise it does not work. 
- For the MPI specification, can we find it automatically? and then give the user's the option to set it by default. Otherwise, it works for me.

af-ayala commented:
> It is working on my end.
> 
> Thank you Gery for doing that.
> 
> Just as a record, when I did cmake without fftw in the PATH, the cmake did not complain.

Sebastien, 
It did not complain since all libraries are off, so your build has all libraries disabled. Once you begin testing a real library, you will need to specify the dependencies.
Let us know if you need something else coded for your convenience.
Thanks.

af-ayala commented on CMakeLists.txt:
```diff
@@ -24,61 +27,20 @@ endif ()
 
 find_package(MPI REQUIRED)
 
-# Add include paths
```
Removing these paths,
How do you we link to multiple 3rd-party libraries?

G-Ragghianti replied:
You can give multiple directories separated by ";".

G-Ragghianti replied:
Separate by semi-colon

af-ayala commented on CMakeLists.txt:
```diff
@@ -24,61 +27,20 @@ endif ()
+link_directories(${FIBER_FFT_LIB_DIR})
 
-if (FIBER_ENABLE_FFTE)
-add_definitions(-DFIBER_ENABLE_FFTE)
```
add_definitions removed does not transfers the def flags required by local files. 
Test this, and it there is another way, then let's fix it.

G-Ragghianti replied:
According to the following documentation, the -D... flags are automatically added to COMPILE_DEFINITIONS.  I have verified that this is the case (using heffte). https://cmake.org/cmake/help/latest/command/add_definitions.html

Are you saying that there is something else that I've missed?

G-Ragghianti replied:
I did test it.  It does automatically propagate the flags.

luszczek commented on CMakeLists.txt:
```diff
@@ -24,61 +27,20 @@ endif ()
-endif()
+# Location of the FFT library
+include_directories(${FIBER_FFT_INCLUDE_DIR})
+link_directories(${FIBER_FFT_LIB_DIR})
```
To link multiple libraries, sometimes multiple directories are required like this: `-DFIBER_FFT_LIB_DIR="/path/to/bheffte;/path/to/ffte-7.0"` I would suggest to rename it to to `_DIRS` or `_DIRECTORIES` suffix. Both are used across the CMake universe but the latter seems to be more closer to the canon.

G-Ragghianti replied:
Yes, we would be able to specify multiple directories this way.  I agree that using "...DIRS" would make this more clear.  Obviously, we would also prefer to be able to automatically search of the libraries, but I this this is preferable to using hard-coded paths or to using separate flags for every library.

luszczek commented on CMakeLists.txt:
```diff
@@ -24,61 +27,20 @@ endif ()
-link_directories($ENV{HOME}/Benchmarks_FFT/p3dfft/build)
-endif()
+# Location of the FFT library
+include_directories(${FIBER_FFT_INCLUDE_DIR})
```
Consider multiple directories (see below).

G-Ragghianti replied:
Agreed.

luszczek commented on CMakeLists.txt:
```diff
@@ -24,61 +27,20 @@ endif ()
-endif()
-
-if (FIBER_ENABLE_FFTMPI)
-add_definitions(-DFIBER_ENABLE_FFTMPI)
```
`add_definitions()` adds a compile-time flag. Now the flag is missing. Is this intended?

G-Ragghianti replied:
According to the following documentation, the -D... flags are automatically added to COMPILE_DEFINITIONS.  I have verified that this is the case (using heffte). https://cmake.org/cmake/help/latest/command/add_definitions.html

luszczek replied:
I think this requires `configure_file()` call and suitable config file. I did it for http://github.com/icl-utk-edu/hpft

G-Ragghianti replied:
Why you make this comment so many times? ;-)

luszczek commented on CMakeLists.txt:
```diff
@@ -24,61 +27,20 @@ endif ()
-endif()
-
-if (FIBER_ENABLE_ACCFFT)
-add_definitions(-DFIBER_ENABLE_ACCFFT)
```
`add_definitions()` adds a compile-time flag. Now the flag is missing. Is this intended?

luszczek commented on CMakeLists.txt:
```diff
@@ -24,61 +27,20 @@ endif ()
-endif()
-
-if (FIBER_ENABLE_P3DFFT)
-add_definitions(-DFIBER_ENABLE_P3DFFT)
```
`add_definitions()` adds a compile-time flag. Now the flag is missing. Is this intended?

luszczek commented on CMakeLists.txt:
```diff
@@ -24,61 +27,20 @@ endif ()
+link_directories(${FIBER_FFT_LIB_DIR})
 
-if (FIBER_ENABLE_FFTE)
-add_definitions(-DFIBER_ENABLE_FFTE)
```
`add_definitions()` adds a compile-time flag. Now the flag is missing. Is this intended?

luszczek commented on CMakeLists.txt:
```diff
@@ -24,61 +27,20 @@ endif ()
-endif()
-
-if (FIBER_ENABLE_SWFFT)
-add_definitions(-DFIBER_ENABLE_SWFFT)
```
`add_definitions()` adds a compile-time flag. Now the flag is missing. Is this intended?

luszczek commented on CMakeLists.txt:
```diff
@@ -24,61 +27,20 @@ endif ()
-endif()
-
-if (FIBER_ENABLE_2DECOMP)
-add_definitions(-DFIBER_ENABLE_2DECOMP)
```
`add_definitions()` adds a compile-time flag. Now the flag is missing. Is this intended?

luszczek commented on CMakeLists.txt:
```diff
@@ -24,61 +27,20 @@ endif ()
-endif()
-
-if (FIBER_ENABLE_NB3DFFT)
-add_definitions(-DFIBER_ENABLE_NB3DFFT)
```
`add_definitions()` adds a compile-time flag. Now the flag is missing. Is this intended?

luszczek reviewed (commented):
Please some comments in the code. I think some functionality is missing after the deletions and the variable names could be improved.

G-Ragghianti commented:
>     1. **Important:**  We need the definitions for enabling every of the 9 libraries, you have removed them all, so it is not linking to any library, the solely reason why it is compiling is that I created shell functions in case people don't have a given library and still tries to use it. Please, at least try it with heFFTe installed, so that you can be sure that other libraries shall not have any issue.

I think that you are assuming this is true without testing my changes.  I have tested it with heffte and it compiles cleanly.  Teh documentation for add_definitions() says that the addition of -D... flags is automatic, and I find that to be true.  If it doesn't work for you, maybe you are using an older cmake?  I was using cmake/3.19.4.  If it is a matter of the cmake version, then we can specify the version that we need.

G-Ragghianti commented:
>     2\. For the DFIBER_FFT_INCLUDE_DIR, I believe the CMake can set this by default, since the path is dependent of where Fiber is installed.

I think you are mistakenly thinking this is for the location of FIBER.  It is in the "FIBER_..." namespace, but it is the location of the FFT libraries that we want to test.

G-Ragghianti commented:
>     4\. For the MPI specification, can we find it automatically? and then give the user's the option to set it by default. Otherwise, it works for me.

Your code was already using find_package() for MPI, so that would be automatic (it was for me).  My code with MPI_DIR allows the user to also set their own MPI location explicitly as you are asking.

G-Ragghianti commented:
> It did not complain since all libraries are off, so your build has all libraries disabled. Once you begin testing a real library, you will need to specify the dependencies.

I don't know if Seb. tested a specific FFT library, but I tested these changes with HEFFTE.  I'd be happy to test withanother library, but I wasn't sure which would be easy to install.  Do you have a recommendation?

G-Ragghianti committed:
Change flag name to clarity

G-Ragghianti commented:
I pushed a change of DIR to DIRS at Piotr recommendation.

One other comment: I recommend that we do not check in any changes that have hard-coded paths or make assumptions about where things are located or how the system is configured.  I know it is tempting when you just want to get something working, but if it only works on a specific machine it isn't much good for anyone else.  Also, someone is going to have to go in and fix it later, so it is better to wait until is it usable by other people before checking it in.   

To avoid any of these kinds of assumptions or unnecessary requirements, I prefer to do a final test in a container.  Otherwise, you will be making something that only builds on saturn, and that looks rather amateur from an outside perspective.

af-ayala commented:
> > ```
> > 1. **Important:**  We need the definitions for enabling every of the 9 libraries, you have removed them all, so it is not linking to any library, the solely reason why it is compiling is that I created shell functions in case people don't have a given library and still tries to use it. Please, at least try it with heFFTe installed, so that you can be sure that other libraries shall not have any issue.
> > ```
> 
> I think that you are assuming this is true without testing my changes. I have tested it with heffte and it compiles cleanly. Teh documentation for add_definitions() says that the addition of -D... flags is automatic, and I find that to be true. If it doesn't work for you, maybe you are using an older cmake? I was using cmake/3.19.4. If it is a matter of the cmake version, then we can specify the version that we need.

Gerald,
Now you have hard coded 
add_definitions(-DFIBER_ENABLE_HEFFTE),
Which before you said it was going to be automatically enabled.
I have tried with recent CMAKE versions and your previous PR was not building the engines appropriately.
I believe the add_definitions need to stay as in the version I put it, unless you know of a way to do it in a more elegant way but still leaving all engines (backends) functional.
We all are able to compile, that is not the issue here, if there was a CI enabled the previous PR would not pass since the heFFTe backend run only its shell functions (i.e. it is not receiving the right definitions. 
Compare your output of:
mpirun -np 2 ./test3D_CPU_R2C heffte
with the one I sent by email to verify correctness. 

Now, if we take your hard code add_definitions(-DFIBER_ENABLE_HEFFTE), as you are doing in your update, when enabling other libraries, we will go back to the same version we have right now in the current CMakeLists. Hard-coding for all libraries.

I am still building a basic CI, but at least, let us try to verify with heffte on our PRs, since we all know heFFTe and have it ready to use in Saturn.

luszczek commented:
I've just ran `cmake -DFIBER_ENABLE_HEFFTE=ON`
`grep -ir dfiber` and got nothing with CMake 19

The old build gave me this:
```
./CMakeFiles/test3D_CPU_R2C.dir/flags.make:C_DEFINES = -DFIBER_ENABLE_HEFFTE
./CMakeFiles/test3D_CPU_C2C.dir/flags.make:C_DEFINES = -DFIBER_ENABLE_HEFFTE
```
I think you need `configure_file()` to propagate definitions. Otherwise I can imagine a lot definitions leaking through.

af-ayala commented on CMakeLists.txt:
```diff
@@ -42,6 +42,8 @@ include_directories(include)
 include_directories(${MPI_DIR}/include)
 link_directories(${MPI_DIR}/lib)
 
+add_definitions(-DFIBER_ENABLE_HEFFTE)
```
**This is a major issue**, if user don't have heFFTe installed and want to try, e.g., FFTE and AccFFT, this will not compile.
That is why the add_definitions in the current CMake are conditionally enable according to the cmake line that user give.

G-Ragghianti replied:
That line was accidentally left in from when I was testing, and I've removed it.

I added back the add_definitions() since the documentation on automatic setting of those flags doesn't seem to be working.  I think it may be because it doesn't propagate the definitions to sub-directories.  We can refine it further , but it works for now.

G-Ragghianti committed:
Adding conditional add_definitions() back in

G-Ragghianti commented:
> add_definitions(-DFIBER_ENABLE_HEFFTE)

This was not intended to be checked in.  I've removed it and added back all the add_definitions().  

And example session that I used for testing is as follows:

```
export HEFFTE=`spack location -i heffte`
cmake -DFIBER_ENABLE_HEFFTE=ON -DFIBER_FFT_INCLUDE_DIRS=$HEFFTE/include -DFIBER_FFT_LIB_DIRS=$HEFFTE/lib ..
make
```

af-ayala commented:
> > add_definitions(-DFIBER_ENABLE_HEFFTE)
> 
> This was not intended to be checked in. I've removed it and added back all the add_definitions().
> 
> And example session that I used for testing is as follows:
> 
> ```
> export HEFFTE=`spack location -i heffte`
> cmake -DFIBER_ENABLE_HEFFTE=ON -DFIBER_FFT_INCLUDE_DIRS=$HEFFTE/include -DFIBER_FFT_LIB_DIRS=$HEFFTE/lib ..
> make
> ```

Thank you for the example session.
Once all 9 libraries are built, we will need to turn ON the ones we want to benchmark, e.g., 8 of the 9.
In the previous cmake, we did it like -DFIBER_ENABLE_FFTE=ON -DFIBER_ENABLE_HEFFTE=ON ....
and so on,

Can you please put an example how to install multiple libraries with your proposed version?
since there is only one parameter (DFIBER_FFT_LIB_DIRS), I am not familiar with how would this work when we need multiple libraries.

luszczek commented:
Just use CMake's list syntax: `-DFIBER_FFT_LIB_DIR=/path/to/heffte;/path/to/fftw;/path/to/ffte`
